Detect False Zero Gas Buildings & Flag Them

vkoves opened the pull request:
# Description

In preparation for our grading work (PR #140 and PR #159 showing grades), added a new anomaly detection step to our data pipeline, and created a new `DataAnomalies` column in `building-benchmarks.csv`, which can be populated with tokens for anomalous data. Thanks to help from jgoldman, there are two flags:

1. **False Gas Use** - flagged in the CSV as `gas:zero-with-prev-use`, which indicates that we detected a building had 0 gas use for the latest year, but used gas in the past, which likely indicates the latest year is a reporting error.
2. **Large Gas Swings** - flagged in the CSV as `gas:large-swings`, this indicates a building had really large shifts in its gas use (like doubling or being cut in half) likely indicating a reporting error. 

Here's a demo with a Shop & Save location:

| Before | After |
| --- | --- |
| ![Screenshot from 2025-02-03 22-36-51](https://github.com/user-attachments/assets/22e194fd-8930-4001-b959-db7acfb7efda) | ![Screenshot from 2025-02-03 22-36-39](https://github.com/user-attachments/assets/b5645ec7-ddd1-433c-aedf-1dad231872ce) |

This catches both buildings tagged in the issue, 10 S State Street & Moody's Solheim Center.

Fixes #120 

# Testing Instructions

You can check the CSV and confirm the only changes are a new column, and confirm that all the buildings marked anomalous indeed have a data issue. The anomaly detection is very simple (and strict), so I'm confident it's accurate. Unit tests exist to validate my work.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

<!-- PR template modified from: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ -->

vkoves commented on run_all.py:
```diff
@@ -0,0 +1,78 @@
+import subprocess
```
I converted my upgrade `run_all` shell script to a Python script, so it's easier to modify and could have unit tests written

vkoves commented on src/data/scripts/add_context_by_property_type.py:
```diff
@@ -28,54 +30,24 @@
 ]
 
-# Columns that should be strings because they are immutable identifiers
-string_cols = [
```
This (and `int_cols`) was duplicated in a lot of places, so I made it a central constant that can be reused.

vkoves commented on src/data/scripts/add_context_by_property_type.py:
```diff
@@ -116,45 +88,61 @@ def calculate_building_stats():
 ### Main
 ###
-def main():
+def main() -> None:
```
Just added some typing and move code into `main()` so i follows best practices, otherwise that code was running on `import`.

vkoves commented:
@jgoldman will take a peek at this and give some feedback while Derek reviews my first PR

jgoldman commented on src/data/scripts/detect_anomalous_buildings.py:
```diff
@@ -0,0 +1,113 @@
+# Output main benchmark path
+input_benchmark_data_csv_path = get_data_file_path(out_dir, 'building-benchmarks.csv')
+
+def detect_anomalous_zero_gas_buildings(historic_data: pd.DataFrame) -> List[int]:
```
You can capture 0 gas buildings as well as emissions that massively spike with this change

```
def detect_anomalous_zero_gas_buildings(historic_data: pd.DataFrame, threshold: float=1.0) -> List[int]:
    """
    Finds buildings in the given historic data DataFrame that sees reported gas use change greater than 100% in a given year in the past, 
    as this is highly likely to be incorrect.  The results are sorted from high to low.

    args:
        historic_data: the dataframe to analyze
        threshold: the percent threshold to detect buildings at (1 = 100%)
    
    returns:
        a list of building IDs that have anomalous gas use
    """

    anom_gas_usage = historic_data.groupby('ID').agg({'NaturalGasUse': determine_abs_delta}).sort_values('NaturalGasUse', ascending=False).dropna().rename(columns={'NaturalGasUse': 'NaturalGasChange'}).reset_index()
    gas_anomaly_ids = anom_gas_usage[anom_gas_usage.NaturalGasChange > threshold]['ID'].unique().tolist()

    print(f"Found {len(gas_anomaly_ids)} buildings with anomalous gas use.\n");

    return gas_anomaly_ids
 ```

jgoldman replied:
Also needs this:
```
def determine_abs_delta(x, n:str = 1):
    shifted_x =  x.shift(1).fillna(x.iloc[0]) + 1
    return abs( (x - shifted_x) / shifted_x).dropna().max()
```

vkoves commented:
@jgoldman - awesome suggestion, thank you for that helpful code! I added that and to keep things clean, I split it out into a separate anomaly output, which now shows like so:

<img src="https://github.com/user-attachments/assets/42dc3aae-5a69-43e1-9ff7-4cd80c0aac4b" width="400">

I then tweaked my previous function to only look for 0 gas use _in the latest year_, since this is the most confusing - as the main tile/number will say 0 (rather than just a blip in the graph where it goes to 0 and rebounds). I updated the anomaly text at the top of the building details to reflect that:

<img src="https://github.com/user-attachments/assets/40b96173-2930-4b15-864b-91be413c1a13" width="400">

I also fixed the stat tile below still celebrating the building was gas free when right above we say it's likely not. The tile now checks for the false gas free data anomaly and flags it:

| Before | After |
| --- | --- |
| ![Screenshot from 2025-02-06 22-03-07](https://github.com/user-attachments/assets/952d5181-2f6e-46aa-8382-0b60d4b63b45) | ![Screenshot from 2025-02-06 22-09-52](https://github.com/user-attachments/assets/b853c7b9-5714-4b1a-b5c0-bfe980c0029e)|

And just to note it, here's the number of anomalies of both types we found:

![Screenshot from 2025-02-06 22-00-09](https://github.com/user-attachments/assets/5634fe95-140d-4ba3-a329-24b8fb2618e5)

jgoldman commented:
Flagging a couple extra items from a quick inspection:

We should potentially leverage reporting status (remove buildings that are not submitted or submitted data)
![image](https://github.com/user-attachments/assets/a32825ef-8ff7-4f2b-9bbb-bbc681903d8f)

This seems like erroneous data since the number of buildings change.  We could check that columns as well
![image](https://github.com/user-attachments/assets/ef3222b4-a537-4ab4-8b36-6d309723d856)

We could build a quick algo that checks that the increase in floor area is not in the last year.  It ignores it if that is the case, otherwise, it flags the building if that area doesn't stay high in subsequent years:

![image](https://github.com/user-attachments/assets/a8e29efb-a63e-4fe3-88a3-78095339311e)
![image](https://github.com/user-attachments/assets/ed9f2a9a-4ee4-4cb5-aaab-c000aedbd1b8)

vkoves commented:
@jgoldman - there's probably a bunch more anomalies we can add in, but for this PR I wanted to start with the biggest and most obvious one I had noticed (fake gas free buildings), which would make any "A" grades we give those buildings look really bad. If you can make a new issue for other anomalies you find, we can then tackle those and figure out if we want to add extra metadata.

For example, some things we might flag but not want to prevent grading for, or some anomalies we might want to mention a specific year (e.g. this building had an anomalous year of data in 2020, but others years looked fine). I think that'll something interesting for us to chew on and figure out how we want to handle

derekeder commented:
@vkoves before I review, should the failing build be addressed?

vkoves commented:
@derekeder - derp, I think I know what that was and my latest commit should fix it, but feel free to review the Python code in the meantime either way, which is the bulk of the changes

vkoves commented:
@derekeder and @jgoldman - I found an interesting quirk I'd love your thoughts on. We flag up [55 E Monroe](https://electrifychicago.net/building/55-e-monroe/) as falsely gas free because it used to have what looks like significant gas usage in the past (> 1 million kBTUs), but in the context of the building, it's actually quite small (< 1%) indicating perhaps a commercial kitchen that got switched to electric or something like that.

I've made this more visible by adding energy mix information to the historical table under "View Extra Technical Info":

| Page Top | Historical Data |
| --- | --- |
| ![Screenshot from 2025-02-14 21-11-47](https://github.com/user-attachments/assets/7fe41be3-2567-4252-ae26-a3987825e855) | ![Screenshot from 2025-02-14 21-11-40](https://github.com/user-attachments/assets/e8979e6d-31f2-4773-817a-9afc908dc0de) |

**The main question - what should we do with this type of building?** I wonder if we should tweak our falsely gas free flag to be something like > 5% gas use or something like that, so that it's not flagging up these buildings that likely didn't have any reporting errors.

derekeder reviewed (commented):
This is a really good improvement to make and flag for users that some data may be reported incorrectly. I bet the city would be interested in seeing this too!

I do think you should have a threshold for 55 E Monroe example. Some buildings will eventually get to zero and the jumps may be significant when they do. 5% seems like a reasonable starting point, but you should continue to spot check.

vkoves commented:
@derekeder - I added a new pre-filter to my gas anomaly detection requiring gas use has to be > 10% of the buildings electric use. This fixed all of the big buildings on the gas free buildings page that I manually wouldn't flag:

![Screenshot from 2025-02-22 13-32-20](https://github.com/user-attachments/assets/36e856bb-9909-4d33-9043-b955318b44cb)

I've updated the tests too to catch this, and I'd love any thoughts on whether that threshold looks good or should get raised more!
